File: 768-partition-labels/partition-labels.java
class Solution {
    public List<Integer> partitionLabels(String s) {
      int n = s.length();
      Map<Character,Integer> map = new HashMap<>();

      for(int i = 0 ; i < n ; i++){
        map.put(s.charAt(i) , i);
      }

      int maxOccurence = 0;
      List<Integer> res = new ArrayList<>();
      int start = 0;

      for(int i = 0 ; i < n ; i++){
        maxOccurence = Math.max(maxOccurence , map.get(s.charAt(i)));

        if(i == maxOccurence){
          res.add(i + 1 - start);
          start = i + 1;
        }
      }

      return res;


    }
}

File: 768-partition-labels/partition-labels.py
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        
        m = {}

        for i,c in enumerate(s):
            m[c] = i
        
        i = 0
        j = 0
        farthest = 0
        res = []

        while j < len(s):
            c = s[j]
            farthest = max(farthest, m[c])

            if j == farthest:
                res.append(j-i+1)
                i = farthest + 1
            
            j += 1
        
        return res

